src/Propulsion.SqlStreamStore/StreamReader.fs

```diff
@@ -0,0 +1,226 @@
+namespace Propulsion.SqlStreamStore
+
+open System
+open System.Text
+
+open SqlStreamStore
+open SqlStreamStore.Streams
+
+open Serilog
+
+open System.Diagnostics
+open Propulsion.Streams
+
+[<AutoOpen>]
+module private Internal =
+
+    [<NoComparison;NoEquality>]
+    type InternalBatch =
+        {
+            firstPosition  : int64
+            lastPosition   : int64
+            messages       : StreamEvent<byte []> array
+            isEnd          : bool
+        }
+        member this.Length = this.messages.Length
+
+    [<RequireQualifiedAccess>]
+    module StreamMessage =
+
+        let intoStreamEvent (msg: StreamMessage) : Propulsion.Streams.StreamEvent<_> =
+            let inline len0ToNull (x : _[]) =
+                match x with
+                | null -> null
+                | x when x.Length = 0 -> null
+                | x -> x
+
+            let data =
+                let json = msg.GetJsonData().Result
+                Encoding.UTF8.GetBytes(json)
+
+            let metadata =
+                Encoding.UTF8.GetBytes(msg.JsonMetadata)
+
+            let event =
+                FsCodec.Core.TimelineEvent.Create (
+                    msg.Position,
```

Comment:
hm, missed this - see other comments
this needs to be the 0-based Index of the event within its original/actual stream, i.e. the `StreamId`
The Index is used to deduplicate/group/order events per stream when buffering
The Position itself is only used for checkpointing purposes (if consuming from CFP or Kafka, a Position is an aggregate value)
(If there was a need to propagate this context inward to a Handler, the TimelineEvent's `Context` property can be used to carry such information, but that has not been done elsewhere)

Reply:
Good catch, changing that should not be a problem.

Reply:
But as far as deduplication and ordering goes, global position should work just fine anyway? Unless it actually is looking for gaps in the indices.

Reply:
The idea is that the StreamsScheduler's buffer holds stuff as arrays per stream in order to deal with incomplete replays

e.g. if kafka input has (for a given stream) 1,2,3 ... 1, 2 ... 3,4 

then the handler is called with contiguous spans with duplicates dropped
so it will typically. be called with [1,2,3] or [1,2,3,4]

I _think_ you're correct that it would actually "work", but each handler invocation would be with a single event only (and anyone depending on Index to mean what it's supposed to mean (ordinal number within the stream) would have issues)